Content.Client/Actions/ActionsSystem.cs

```diff
@@ -64,12 +110,172 @@ public override void Initialize()
-                .BindBefore(EngineKeyFunctions.Use, new PointerInputCmdHandler(TargetingOnUse),
+                .BindBefore(EngineKeyFunctions.Use, new PointerInputCmdHandler(TargetingOnUse, outsidePrediction: true),
                     typeof(ConstructionSystem), typeof(DragDropSystem))
+                .BindBefore(EngineKeyFunctions.UIRightClick, new PointerInputCmdHandler(TargetingCancel, outsidePrediction: true))
                 .Register<ActionsSystem>();
 
-            SubscribeLocalEvent<ClientActionsComponent, PlayerAttachedEvent>((_, component, _) => component.PlayerAttached());
-            SubscribeLocalEvent<ClientActionsComponent, PlayerDetachedEvent>((_, component, _) => component.PlayerDetached());
+            SubscribeLocalEvent<ActionsComponent, PlayerAttachedEvent>(OnPlayerAttached);
+            SubscribeLocalEvent<ActionsComponent, PlayerDetachedEvent>(OnPlayerDetached);
+            SubscribeLocalEvent<ActionsComponent, ComponentHandleState>(HandleState);
+        }
+
+        protected override void Dirty(ActionType action)
+        {
+            base.Dirty(action);
+            if (action.AttachedEntity == _playerManager.LocalPlayer?.ControlledEntity)
+                UIDirty = true;
+        }
+
+        private void HandleState(EntityUid uid, ActionsComponent component, ref ComponentHandleState args)
+        {
+            if (args.Current is not ActionsComponentState state)
+                return;
+
+            var serverActions = new SortedSet<ActionType>(state.Actions);
+
+            foreach (var act in component.Actions.ToList())
+            {
+                if (act.ClientExclusive)
+                    continue;
+
+                if (!serverActions.TryGetValue(act, out var serverAct))
+                {
+                    component.Actions.Remove(act);
+                    continue;
+                }
+
+                act.CopyFrom(serverAct);
+                serverActions.Remove(serverAct);
+
+                if (act is EntityTargetAction entAct)
+                {
+                    entAct.Whitelist?.UpdateRegistrations();
+                }
+            }
+
+            // Anything that remains is a new action
+            foreach (var newAct in serverActions)
+            {
+                if (newAct is EntityTargetAction entAct)
+                    entAct.Whitelist?.UpdateRegistrations();
+
+                // We create a new action, not just sorting a reference to the state's action.
+                component.Actions.Add((ActionType) newAct.Clone());
+            }
+
+            UIDirty = true;
+        }
+
+        /// <summary>
+        /// Highlights the item slot (inventory or hand) that contains this item
+        /// </summary>
+        /// <param name="item"></param>
+        public void HighlightItemSlot(EntityUid item)
+        {
+            StopHighlightingItemSlots();
+
+            _highlightedEntity = item;
+            _itemSlotManager.HighlightEntity(item);
+        }
```

Comment:
Maybe account for highlighting more than one entity. Assuming itemslotman supports this.

Reply:
The manager does support highlighting more than one slot at a time, but AFAIK actions wouldn't need to use that. The highlighting is just used to highlight the item that provides/enables a given action, which is only ever one item. `StopHighlightingItemSlots()` only removes the highlight from the item that was previously highlighted, and doesn't clear other active highlights. 

I am going to rename `StopHighlightingItemSlots` to be singular (Slots -> Slot), and make `_highlightedEntity` nullable (currently uses default/invalid Uid).